docs/proposals/19-migrating-to-prometheus-operator.md

```diff
@@ -0,0 +1,529 @@
+be more difficult to give Prometheus invalid config.
+
+### Goals
+
+- Migrate from the current monitoring stack to the prometheus-operator.
+
+- Improve the monitoring extensibility and improve developer experience when
+  adding or editing configuration. This includes the monitoring extensions in
+  addition to core Gardener components.
+
+- Provide a clear direction on how monitoring resources should be managed.
+  Currently, there are many ways that monitoring configuration is being deployed
+  and this should be unified.
+
+- Improve how dashboards are discovered and provisioned for Grafana. Currently,
+  all dashboards are appended into a single configmap. This can be an issue if
+  the maximum configmap size of 1MiB is ever exceeded.
+
+### Non-Goals
+
+- Changes to the logging stack.
+
+- Feature parity between the current solution and the one proposed in this GEP.
+  The new stack should provide similar monitoring coverage, but it will be very
+  difficult to evaluate if there is feature parity. Perhaps some features will
+  be missing, but others may be added.
+
+## Proposal
+
+Today, Gardener provides monitoring for shoot clusters (i.e. system components
+and the control plane) and for the seed cluster. The proposal is to migrate
+these monitoring stacks to use the [prometheus-operator]. The proposal is lined
+out below:
+
+### API
+
+Use the [API] provided by the [prometheus-operator] and create Go structs.
+
+### Prometheus Operator CRDs
+
+Deploy the [prometheus-operator] and its CRDs. These components can be deployed
+via `ManagedResources`. The operator itself and some other components outlined
+in the GEP will be deployed in a new namespace called `monitoring`. The CRDs for
+the [prometheus-operator] and the operator itself can be found
+[here][prom-crd-bundle]. This step is a prerequisite for all other steps.
+
+### Shoot Monitoring
+
+Gardener will create a monitoring stack similar to the current one with the
+[prometheus-operator] custom resources.
+
+1. Most of the shoot monitoring is deployed via this
+    [chart][shoot-monitoring]. The goal is to create a similar stack, but not
+    necessarily with feature parity, using the [prometheus-operator].
+
+    - An example Prometheus object that would be deployed in a shoot's
+      control plane.
+
+    ```yaml
+    apiVersion: monitoring.coreos.com/v1
+    kind: Prometheus
+    metadata:
+      labels:
+        app: prometheus
+      name: prometheus
+      namespace: shoot--project--name
+    spec:
+      enableAdminAPI: false
+      logFormat: logfmt
+      logLevel: info
+      image: image:tag
+      paused: false
+      portName: web
+      replicas: 1
+      retention: 30d
+      routePrefix: /
+      serviceAccountName: prometheus
+      serviceMonitorNamespaceSelector:
+        matchExpressions:
+        - key: kubernetes.io/metadata.name
+          operator: In
+          values:
+          - monitoring
```

Comment:
```suggestion
```

The *Monitors are going to be only in the shoot control plane namespace.

Even simpler alternative: leave the namespace selector to be nil.

> Namespace’s labels to match for ServiceMonitor discovery. If nil, only check own namespace.

Note that there are 2 more similar occurrences further below.